### ISSUE TITLE
fmodule-ts parameter caused an exception

### ISSUE THREAD

**Issue by Season0518** [iains/gcc-14-branch#17]
The issue has been reflected at clangd/clangd#2243, and I believe it is more likely to be caused by gcc.

How to reproduce?
 - Create an file: `empty.cpp`
 ```
#include <cassert>
int main(){
}
```
 - Use compilation instructions:
 ```
g++-14  -g -std=c++23 -arch arm64 - isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX15.2.sdk -fmodules-ts empty.cpp
```

You will see that the program error and the error thrown by clangd are basically the same:
![image](https://github.com/user-attachments/assets/8e245622-00b0-4dab-bd7d-45ab78c4a9bb)

But if we follow the solution at clangd/clangd#2243 and downgrade to c++17 or below (disable modules features), the program will compile normally and there will be no errors.
Corresponding compilation instructions (remove module features)

```
g++-14  -g -std=c++23 -arch arm64 - isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX15.2.sdk empty.cpp
```

![image](https://github.com/user-attachments/assets/5668ac8f-dbb0-48d2-a998-762f016061a2)

SDK:
xcode-select: version 2409
gcc: gcc-13/gcc-12/gcc-14 (Homebrew GCC 13.2.0) 14.2.0
clangd: Homebrew clangd version 19.1.4

Operating system:

MacOS version: 15.1.1 (24B91)

**Comment by iains** [comment]
this seems odd
`gcc: gcc-13/gcc-12/gcc-14 (Homebrew GCC 13.2.0) 14.2.0`

* is it 13 or 14?

* was the gcc-14 built against the 15.2 SDK?

In general, please ask about homebrew problems there first - since I am not involved in the details of their distributions.

**Comment by Season0518** [comment]
> this seems odd `gcc: gcc-13/gcc-12/gcc-14 (Homebrew GCC 13.2.0) 14.2.0`
> 
> * is it 13 or 14?

I'm sorry for misleading you. This stuff was copied from the previous issue.In this issue, the instruction uses **gcc-14**. Other versions seem to be able to reproduce the issue as well.

> * was the gcc-14 built against the 15.2 SDK?

may be 15.2。

compiler verbose:

```
g++-14 -v
Using built-in specs.
COLLECT_GCC=g++-14
COLLECT_LTO_WRAPPER=/opt/homebrew/Cellar/gcc/14.2.0_1/bin/../libexec/gcc/aarch64-apple-darwin24/14/lto-wrapper
Target: aarch64-apple-darwin24
Configured with: ../configure --prefix=/opt/homebrew/opt/gcc --libdir=/opt/homebrew/opt/gcc/lib/gcc/current --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran,m2 --program-suffix=-14 --with-gmp=/opt/homebrew/opt/gmp --with-mpfr=/opt/homebrew/opt/mpfr --with-mpc=/opt/homebrew/opt/libmpc --with-isl=/opt/homebrew/opt/isl --with-zstd=/opt/homebrew/opt/zstd --with-pkgversion='Homebrew GCC 14.2.0_1' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --with-system-zlib --build=aarch64-apple-darwin24 --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX15.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 14.2.0 (Homebrew GCC 14.2.0_1) 
```

**Comment by iains** [comment]
notes:
1. modules support is (very) experimental in GCC-14 - although many fixes have been made for the upcoming GCC-15, so it might be better just to avoid the problem case(s)
2. there are no modules-related changes on the Darwin branches .. so I will see about how to reproduce this on the "upstream".

**Comment by Season0518** [comment]
> 1. modules support is (very) experimental in GCC-14 - although many fixes have been made for the upcoming GCC-15, so it might be better just to avoid the problem case(s)

Roger.

> 2. there are no modules-related changes on the Darwin branches .. so I will see about how to reproduce this on the "upstream".

But based on the phenomenon I have observed(Clangd throw error and compile with specific flags), it seems that the problem is caused by a system(CommandLine Tools) upgrade,It's so weird.

**Comment by iains** [comment]
> > 2. there are no modules-related changes on the Darwin branches .. so I will see about how to reproduce this on the "upstream".
> 
> But based on the phenomenon I have observed(Clangd throw error and compile with specific flags), it seems that the problem is caused by a system(CommandLine Tools) upgrade,It's so weird.

* maybe not so weird - The SDK (and therefore the headers) also changes between tools updates; e.g. it could be that  `MacOSX15.sdk` points to 15.1 in one release and 15.2 in the next.

* OK .. I also get some similar kinds of errors on gcc-14.2darwin-release2 + x86_64-darwin23 + SDK 15.1 .. so now I need to take a look at what might be causing this (and also if it repeats on unpatched upstream..

**Comment by iains** [comment]
the problem repeats (for x86_64) with unpatched releases/gcc-14 so this needs handling upstream - it appears not related to the changes on the Darwin development branch.

**Comment by Season0518** [comment]
> the problem repeats (for x86_64) with unpatched releases/gcc-14

Yes,It's a common problem. It's also can be reproduced on my Apple Silicon Mac

> it appears not related to the changes on the Darwin development branch.

I'm sorry to hear that.

However, can we consider the following two approaches in GCC-14 to ensure the proper functioning of LSP (Clangd)?
 - Reduce dependence on Command Line Tools (Related: clangd/vscode-clangd#473)
 - When LSPs such as clangd request the compile flag from gcc, the modules feature is disabled by default.

The above are my guesses based on logs and outputs, as I do not have a deep understanding of toolchains.I hope my guess is correct

**Comment by iains** [comment]
if you follow the suggestions in the diagnostics, for example:
```
note: ‘ptrdiff_t’ is defined in header ‘<cstddef>’; this is probably fixable by adding ‘#include <cstddef>’
```

this works for me with 15.1 ..

```
#include <cstddef>
#include <cassert>

int main ()
{}

```

Does that work correctly for you ?

**Comment by iains** [comment]
For the record, (on x86_64) the issue is triggered by the following code in  `MacOSX15.1.sdk/usr/include/i386/_types.h`

```c++
#if defined(__has_feature) && __has_feature(modules)
#define USE_CLANG_TYPES 1
#else
#define USE_CLANG_TYPES 0
#endif
```

Which causes some additional includes to be used - which reference the undefined entities defined in stddef.

I don't quite know if this is an inconsistency between `clang modules` and `c++2x modules` or if there's a more general issue.

Anyway - including `stddef` should deal with at least this set.

@fxcoudert have you encountered this anywhere yet?

**Comment by Season0518** [comment]
> if you follow the suggestions in the diagnostics

Yes, that does fix the problem, thank you very much.

But I am curious (although not so relevant to this repository) why clangd still throws the same type of error?

![image](https://github.com/user-attachments/assets/dd3cdbdd-5c81-4832-b915-89372452e549)

**Comment by fxcoudert** [comment]
@iains yes I've encountered that `USE_CLANG_TYPES` issue once before. It is causing this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=116827

**Comment by iains** [comment]
> @iains yes I've encountered that `USE_CLANG_TYPES` issue once before. It is causing this: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=116827

hmmm... this might not be fixable without a fixincludes :( ...

**Comment by iains** [comment]
please try the latest branch version, https://github.com/iains/gcc-14-branch/commit/50f92929f337c81aaf330bb04ac0675dba340e25 is a proposed fix for this.
